src/storage/recovery/recovery_manager.cpp

```diff
@@ -0,0 +1,934 @@
+                    table_pr->AccessWithNullCheck(pr_map[oid]), size);
+      }
+    }
+
+    if (insert) {
+      bool result UNUSED_ATTRIBUTE = (index->GetConstraintType() == index::ConstraintType::UNIQUE)
+                                         ? index->InsertUnique(txn, *index_pr, tuple_slot)
+                                         : index->Insert(txn, *index_pr, tuple_slot);
+      TERRIER_ASSERT(result, "Insert into index should always succeed for a committed transaction");
+    } else {
+      index->Delete(txn, *index_pr, tuple_slot);
+    }
+  }
+
+  delete[] index_buffer;
+}
+
+uint32_t RecoveryManager::ProcessSpecialCaseCatalogRecord(
+    transaction::TransactionContext *txn, std::vector<std::pair<LogRecord *, std::vector<byte *>>> *buffered_changes,
+    uint32_t start_idx) {
+  auto *curr_record = buffered_changes->at(start_idx).first;
+  TERRIER_ASSERT(curr_record->RecordType() == LogRecordType::REDO || curr_record->RecordType() == LogRecordType::DELETE,
+                 "Special case catalog record must be redo or delete");
+
+  // Get oid of table this record modifies
+  catalog::table_oid_t table_oid;
+  if (curr_record->RecordType() == LogRecordType::REDO) {
+    table_oid = curr_record->GetUnderlyingRecordBodyAs<RedoRecord>()->GetTableOid();
+  } else {
+    table_oid = curr_record->GetUnderlyingRecordBodyAs<DeleteRecord>()->GetTableOid();
+  }
+
+  switch (!table_oid) {
+    case (!catalog::DATABASE_TABLE_OID): {
+      return ProcessSpecialCasePGDatabaseRecord(txn, buffered_changes, start_idx);
+    }
+
+    case (!catalog::CLASS_TABLE_OID): {
+      return ProcessSpecialCasePGClassRecord(txn, buffered_changes, start_idx);
+    }
+
+    case (!catalog::COLUMN_TABLE_OID): {
+      TERRIER_ASSERT(curr_record->RecordType() == LogRecordType::DELETE,
+                     "Special case pg_attribute record must be a delete");
+      // A delete into pg_attribute means we are deleting a column. There are two cases:
+      //  1. Drop column: This requires some additional processing to actually clean up the column
+      //  2. Cascading delete from drop table: In this case, we don't process the record because the DeleteTable catalog
+      //  function will clean up the columns
+      // We currently don't support Case 1
+      return 0;  // Case 2, no additional records processed
+    }
+
+    case (!catalog::INDEX_TABLE_OID): {
+      TERRIER_ASSERT(curr_record->RecordType() == LogRecordType::DELETE,
+                     "Special case pg_index record must be a delete");
+      // A delete into pg_index means we are dropping an index. In this case, we dont process the record because the
+      // DeleteIndex catalog function will cleanup the entry in pg_index for us.
+      return 0;  // No additional logs processed
+    }
+
+    default:
+      throw std::runtime_error("This table does not require special case logic for record processing");
```

Comment:
Should this be a runtime error?  I think you've already screened with `IsSpecialCaseCatalogRecord` so presumably hitting this means we messed up in our logic and not an input error.